### PR TITLE
LASolver: Add cuts-from-proofs algorithm

### PR DESCRIPTION
**Pull request by blishko**
Cuts-from-proofs is a sound and complete technique for solving linear integer arithmetic.
It aims at removing whole faces of rational polygon of solutions without any integer point by proving that a set of integer equalities does not have a solution. These equalities are taken from the initial set of inequalities by identifying *defining constraints*, which are the the constraints satisfied by current rational solution *tightly*.

The proof of unsatisfiability is a constraint that is a linear combination of defining constraints and is unsatisfiable. We can then branch around the plane of this constraint to exclude this reason of unsatisfiability and steer the simplex into more promising regions.

### CODE REVIEW

File: src/tsolvers/lasolver/CutCreator.cc [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+void ColMatrix::Col::negate() {
+    this->poly.negate();
+}
+
+void ColMatrix::Col::add(Col const & other, opensmt::Real const & multiple) {
+    this->poly.merge(other.poly, multiple, [](LVRef){}, [](LVRef){});
+}
+
+opensmt::Real ColMatrix::Col::product(const std::vector<FastRational> & values) const {
+    opensmt::Real sum = 0;
```
[the remark this answers is not in the document]
Comment (blishko) [reply src/tsolvers/lasolver/CutCreator.cc:21]: The number of elements in the column does not matter, unless we do not have `value` for each row. But there is an `assert` for that below, in the `for` loop.

[the remark this answers is not in the document]
Comment (aehyvari) [reply src/tsolvers/lasolver/CutCreator.cc:21]: I know.  It's just surprised me that I couldn't write an assertion like that.

File: src/tsolvers/lasolver/CutCreator.cc [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+PTRef ColMatrix::Col::buildCutConstraint(const std::vector<PTRef> & toVarMap, ArithLogic & logic) const {
+    vec<PTRef> args; args.capacity(poly.size());
+    for (auto const & term : poly) {
+        args.push(logic.mkTimes(toVarMap[term.var.x], logic.mkIntConst(term.coeff)));
+    }
+    return logic.mkPlus(std::move(args));
+}
+
+const FastRational * ColMatrix::Col::tryGetCoeffFor(RowIndex rowIndex) const {
+    LVRef bound {static_cast<uint32_t>(rowIndex.count)};
```
[the remark this answers is not in the document]
Comment (aehyvari) [reply src/tsolvers/lasolver/CutCreator.cc:39]: I'll work this out in the PR.

**Comment by blishko** [comment]
OK, I will merge this and you can create your PR at any point.
I want to re-run the experiments for MBTC on top of this, I would like to merge #388 next.

**Pull request merged**